aiida_quantumespresso/parsers/pp.py

```diff
@@ -352,34 +357,31 @@ def parse_gaussian(self, data_file_str):
         voxel_array = np.array([[x_line[1], x_line[2], x_line[3]], [y_line[1], y_line[2], y_line[3]],
                                 [z_line[1], z_line[2], z_line[3]]],
                                dtype=np.float64)
+        atomic_numbers = np.empty(natoms, int)
+        coordinates = np.empty((natoms, 3))
+        for i in range(natoms):
+            line = header[6 + i].split()
+            atomic_numbers[i] = int(line[0])
+            coordinates[i] = [float(s) for s in line[2:]]
 
         # Get the volumetric data
-        data_array = np.zeros((xdim, ydim, zdim))
-
-        # The data is organised into columns with a fixed number of values
-        # Gather up all the data points into one list for easier unpacking
-        datalist = []
+        data_array = np.empty(xdim * ydim * zdim, dtype=float)
+        cursor = 0
         for line in data_lines:
-            for i in range(0, len(line), 13):
-                data_point = line[i:i + 13].strip()
-                if data_point != '':
-                    datalist.append(float(data_point))
-
-        # Unpack the list and repack as a 3D array
-        # Note the unusual indexing: cube files run over the z index first, then y and x.
-        # E.g. The first volumetric data point is x,y,z = (0,0,0) and the second is (0,0,1)
-        for i in range(0, xdim):
-            for j in range(0, ydim):
-                for k in range(0, zdim):
-                    data_array[i, j, k] = (datalist[(i * ydim * zdim) + (j * zdim) + k])
+            ls = line.split()
+            data_array[cursor:cursor + len(ls)] = ls
+            cursor += len(ls)
+        data_array = data_array.reshape((xdim, ydim, zdim))
 
         coordinates_units = 'bohr'
         data_units = self.units_dict[self.output_parameters['plot_num']]
 
         arraydata = orm.ArrayData()
         arraydata.set_array('voxel', voxel_array)
         arraydata.set_array('data', data_array)
-        arraydata.set_array('coordinates_units', np.array(coordinates_units))
         arraydata.set_array('data_units', np.array(data_units))
+        arraydata.set_array('coordinates_units', np.array(coordinates_units))
+        arraydata.set_array('coordinates', coordinates)
```

Comment:
why are we adding this? These are the coordinates of the atoms? This is redundant information, because it can be retrieved from the structure that was input of the parent calculation

Reply:
In nanoribbon work chain we reduce the size of the cube files before parsing them, as they contain a lot of data points with values close to zero. This reduction involves changes in the structure's positions.

> why are we adding this? This is redundant information ...

In the view what I wrote above it would be nice for the standard parser to be able to extract the coordinates so that we don't need to write our own parser for that.

Reply:
Wait, how do you reduce the size of the cube files before parsing them? They are generated by `pp.x` and parsed straight away, no? Wouldn't the more reasonable approach be to have `pp.x` generate less points in the first place?

Reply:
> Wait, how do you reduce the size of the cube files before parsing them?

I crop them: https://github.com/nanotech-empa/aiida-nanotech-empa/blob/nanoribon_workchain/aiida_nanotech_empa/parsers.py#L21-L27

> They are generated by pp.x and parsed straight away, no?

I redefine `parse_gaussian` function and crop the file before giving it to the standard `parse_gaussian` 

> Wouldn't the more reasonable approach be to have pp.x generate less points in the first place?

How do you do that? One cannot modify the cell size in the pp.x, which leaves us with changing the density of points as the only option, which leads to a poor data quality. Instead, we take the whole cube file, crop it,  keeping only the nanoribbon's surrounding, and give it to the standard parser afterwords.

Reply:
After a discussion, we agreed to leave this part out for another PR.